lib/admin.rb

```diff
@@ -0,0 +1,90 @@
+require 'date'
+require_relative 'reservation'
+require_relative 'block_room'
+
+class Admin
+  attr_reader :roomlist, :reservations, :price, :blocks
+
+  def initialize
+    @roomlist = []
+    @reservations = []
+    @blocks = []
+
+    20.times do |i|
+      @roomlist << (i + 1)
+    end
+  end
+
+  def add_reservation(check_in, check_out, room_id)
+    available_rooms = available_rooms(check_in, check_out)
+
+    if !(available_rooms.include?(room_id))
+      raise ArgumentError.new("#{room_id} is booked for these dates")
+    end
+
+    if available_rooms.count == 0
+      raise ArgumentError.new("No available roooms")
+    end
+
+    @reservations << Reservation.new(check_in, check_out, room_id)
+  end
+
+  def available_rooms(check_in, check_out)
+    check_in = Date.parse(check_in)
+    check_out = Date.parse(check_out)
+
+    rooms = @roomlist.dup
+
+    @reservations.each do |reservation|
+      valid_check_in = check_in.between?(reservation.start_date, (reservation.end_date - 1))
+
+      valid_check_out = check_out.between?((reservation.start_date + 1), reservation.end_date)
```

Comment:
What about if the reservation occurs entirely between `check_in` and `check_out`?